File: 2025-05-MAY-GFG-POTD/May 06 - Left View of Binary Tree/Solution.cpp
#include <iostream>
#include <sstream>
#include <vector>
#include <queue>
using namespace std;

struct Node {
    int data;
    Node* left;
    Node* right;

    Node(int val) : data(val), left(nullptr), right(nullptr) {}
};

Node* buildTree(string str) {
    if (str.empty() || str[0] == 'N') return nullptr;

    vector<string> ip;
    istringstream iss(str);
    string val;
    while (iss >> val) ip.push_back(val);

    Node* root = new Node(stoi(ip[0]));
    queue<Node*> q;
    q.push(root);

    int i = 1;
    while (!q.empty() && i < ip.size()) {
        Node* currNode = q.front();
        q.pop();

        if (ip[i] != "N") {
            currNode->left = new Node(stoi(ip[i]));
            q.push(currNode->left);
        }
        i++;
        if (i >= ip.size()) break;

        if (ip[i] != "N") {
            currNode->right = new Node(stoi(ip[i]));
            q.push(currNode->right);
        }
        i++;
    }

    return root;
}

vector<int> leftView(Node* root) {
    vector<int> result;
    if (!root) return result;

    queue<Node*> q;
    q.push(root);

    while (!q.empty()) {
        int levelSize = q.size();

        for (int i = 0; i < levelSize; i++) {
            Node* curr = q.front();
            q.pop();

            if (i == 0)
                result.push_back(curr->data);

            if (curr->left)
                q.push(curr->left);
            if (curr->right)
                q.push(curr->right);
        }
    }

    return result;
}

int main() {
    int t;
    cin >> t;
    cin.ignore();
    while (t--) {
        string s;
        getline(cin, s);
        Node* root = buildTree(s);
        vector<int> view = leftView(root);
        for (int val : view) {
            cout << val << " ";
        }
        cout << endl;
        cout << "~" << endl;
    }
    return 0;
}

File: 2025-05-MAY-GFG-POTD/May 06 - Left View of Binary Tree/Solution.py
from collections import deque

class Node:
    def __init__(self, data):
        self.data = data
        self.left = None
        self.right = None

def buildTree(s):
    if not s or s[0] == 'N':
        return None

    ip = s.strip().split()
    root = Node(int(ip[0]))
    q = deque([root])

    i = 1
    while q and i < len(ip):
        curr = q.popleft()

        if ip[i] != 'N':
            curr.left = Node(int(ip[i]))
            q.append(curr.left)
        i += 1
        if i >= len(ip): break

        if ip[i] != 'N':
            curr.right = Node(int(ip[i]))
            q.append(curr.right)
        i += 1

    return root

def leftView(root):
    result = []
    if not root:
        return result

    q = deque([root])

    while q:
        level_size = len(q)
        for i in range(level_size):
            node = q.popleft()
            if i == 0:
                result.append(node.data)
            if node.left:
                q.append(node.left)
            if node.right:
                q.append(node.right)

    return result

# Example usage
if __name__ == "__main__":
    t = int(input())
    for _ in range(t):
        s = input()
        root = buildTree(s)
        result = leftView(root)
        print(' '.join(map(str, result)))
        print("~")

